File: Data Structures & Algorithms/matrixDFS/submission-1.kt
class Solution {

    fun countPaths(grid: Array<IntArray>): Int {
        val rows = grid.size
        val cols = grid[0].size
        val visited = Array(rows) { BooleanArray(cols) }

        val directions = arrayOf(
            intArrayOf(1, 0),
            intArrayOf(-1, 0),
            intArrayOf(0, 1),
            intArrayOf(0, -1)
        )

        fun dfs(row: Int, col: Int): Int {
            if (
                row !in 0 until rows ||
                col !in 0 until cols ||
                visited[row][col] ||
                grid[row][col] == 1
            ) {
                return 0
            }

            if (row == rows - 1 && col == cols - 1) {
                return 1
            }

            visited[row][col] = true

            var paths = 0
            for ((dr, dc) in directions) {
                paths += dfs(row + dr, col + dc)
            }

            visited[row][col] = false

            return paths
        }

        return dfs(0, 0)
    }
}

File: Data Structures & Algorithms/island-perimeter/submission-1.cpp
class Solution {
   public:
    int islandPerimeter(vector<vector<int>>& grid) {
        int rows = grid.size();
        int cols = grid[0].size();

        vector<vector<bool>> visited(rows, vector<bool>(cols, false));

        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                if (grid[row][col] == 1) {
                    return dfs(grid, visited, rows, cols, row, col);
                }
            }
        }

        return 0;
    }

   private:
    int dfs(vector<vector<int>>& grid, vector<vector<bool>>& visited, int rows, int cols, int row,
            int col) {
        // Outside grid means one exposed edge.
        if (row < 0 || row >= rows || col < 0 || col >= cols) {
            return 1;
        }

        // Water means one exposed edge.
        if (grid[row][col] == 0) {
            return 1;
        }

        // Already counted this land cell.
        if (visited[row][col]) {
            return 0;
        }

        visited[row][col] = true;

        int perimeter = 0;
        perimeter += dfs(grid, visited, rows, cols, row, col - 1);  // left
        perimeter += dfs(grid, visited, rows, cols, row - 1, col);  // up
        perimeter += dfs(grid, visited, rows, cols, row, col + 1);  // right
        perimeter += dfs(grid, visited, rows, cols, row + 1, col);  // down

        return perimeter;
    }
};